### PR TITLE
Thrust tutorial

### PR DESCRIPTION
**Pull request by dtmoodie**
Tutorial and helper files for making a thrust style iterator from a GpuMat.

### CODE REVIEW

**Comment by Dikay900** [comment]
The buildbot is ok with this except with the tabs in the files.
http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/4303/steps/whitespace/logs/stdio

Please use 4 whitespace chars instead of a tab then this error will be resolved.

**Comment by Dikay900** [comment]
there is one last trailing whitespace :)
http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/4311/steps/whitespace/logs/stdio

**Comment by vpisarev** [comment]
@jet47, could you please evaluate this tutorial?

File: samples/cpp/tutorial_code/gpu/gpu-thrust-interop/Thrust_interop.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+/*
+    @Brief GpuMatBeginItr returns a thrust compatible iterator to the beginning of a GPU mat's memory.
+    @Param mat is the input matrix
+    @Param channel is the channel of the matrix that the iterator is accessing.  If set to -1, the iterator will access every element in sequential order
+*/
+template<typename T>
+thrust::permutation_iterator<thrust::device_ptr<T>, thrust::transform_iterator<step_functor<T>, thrust::counting_iterator<int>>>  GpuMatBeginItr(cv::cuda::GpuMat mat, int channel = 0)
+{
+    if (channel == -1)
+        mat = mat.reshape(1);
```
Comment (jet47) [inline samples/cpp/tutorial_code/gpu/gpu-thrust-interop/Thrust_interop.hpp]: I guess you should set `channel` to `0` after `mat = mat.reshape(1);`.

**Comment by dtmoodie** [comment]
You're correct.

**Comment by jet47** [comment]
:+1:

**Comment by mshabunin** [comment]
@dtmoodie , could you please squash all commits into one?

**Pull request merged**